libdnf/dnf-sack.c

```diff
@@ -1246,14 +1312,133 @@ void
 dnf_sack_set_includes(DnfSack *sack, DnfPackageSet *pset)
 {
     DnfSackPrivate *priv = GET_PRIVATE(sack);
-    priv->pkg_includes = free_map_fully(priv->pkg_includes);
+    dnf_sack_set_excludes_includes(sack, &priv->pkg_includes, pset);
+}
+
+/**
+ * dnf_sack_reset_excludes:
+ * @sack: a #DnfSack instance.
+ *
+ * Reset excludes (remove excludes map from memory).
+ *
+ * Since: 0.9.4
+ */
+void
+dnf_sack_reset_excludes(DnfSack *sack)
+{
+    dnf_sack_set_excludes(sack, NULL);
+}
+
+/**
+ * dnf_sack_reset_includes:
+ * @sack: a #DnfSack instance.
+ *
+ * Reset includes (remove includes map from memory).
+ *
+ * Since: 0.9.4
+ */
+void
+dnf_sack_reset_includes(DnfSack *sack)
+{
+    dnf_sack_set_includes(sack, NULL);
+}
+
+/**
+ * dnf_sack_get_excludes:
+ * @sack: a #DnfSack instance.
+ * @pset: a #DnfPackageSet or %NULL.
+ *
+ * Gets sack excludes.
+ *
+ * Since: 0.9.4
+ */
+DnfPackageSet *
+dnf_sack_get_excludes(DnfSack *sack)
+{
+    DnfSackPrivate *priv = GET_PRIVATE(sack);
+    Map *excl = priv->pkg_excludes;
+    return excl ? dnf_packageset_from_bitmap(sack, excl) : NULL;
+}
+
+/**
+ * dnf_sack_get_includes:
+ * @sack: a #DnfSack instance.
+ *
+ * Gets sack includes.
+ *
+ * Since: 0.9.4
+ */
+DnfPackageSet *
+dnf_sack_get_includes(DnfSack *sack)
+{
+    DnfSackPrivate *priv = GET_PRIVATE(sack);
+    Map *incl = priv->pkg_includes;
+    return incl ? dnf_packageset_from_bitmap(sack, incl) : NULL;
+}
 
-    if (pset) {
-        Map *nincl = dnf_packageset_get_map(pset);
-        priv->pkg_includes = g_malloc0(sizeof(Map));
-        map_init_clone(priv->pkg_includes, nincl);
+/**
+ * dnf_sack_set_use_includes:
+ * @sack: a #DnfSack instance.
+ * @repo_name: a name of repo or %NULL for all repos.
+ * @enabled: a use includes for a repo or all repos.
+ *
+ * Enable/disable usage of includes for repo/all-repos.
+ *
+ * Returns: FALSE if error occured (unknown reponame) else TRUE.
+ *
+ * Since: 0.9.4
+ */
+gboolean
+dnf_sack_set_use_includes(DnfSack *sack, const char *reponame, gboolean enabled)
+{
+    DnfSackPrivate *priv = GET_PRIVATE(sack);
+    Pool *pool = dnf_sack_get_pool(sack);
+
+    if (reponame) {
+        HyRepo hyrepo = hrepo_by_name(sack, reponame);
+        if (!hyrepo)
+            return FALSE;
+        if (hy_repo_get_use_includes(hyrepo) != enabled)
```

Comment:
Yes. This is optimization. I want run code on line 1404 `priv->considered_uptodate = FALSE;` only if state of repo was changed.